Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/Win32Application.cpp

```diff
@@ -0,0 +1,166 @@
+        nullptr,        // We have no parent window.
+        nullptr,        // We aren't using menus.
+        hInstance,
+        pSample);
+    if (!m_hwnd)
+    {
+        MessageBox(NULL, L"Error creating window",
+            L"Error", MB_OK | MB_ICONERROR);
+        return false;
+    }
+
+    // Initialize the sample and show the window
+    pSample->OnInit();
+    ShowWindow(m_hwnd, nCmdShow);
+
+    // Main sample loop.
+    close_window = false;
+    MSG msg = {};
+    while (msg.message != WM_QUIT)
+    {
+        // Destroy the sample and window if the close_window
+        // variable is set to true
+        if (close_window) {
+            pSample->OnDestroy();
+            PostMessage(m_hwnd, WM_CLOSE, 0, 0);
+            break;
+        }
+        // Process any messages in the queue.
+        if (PeekMessage(&msg, NULL, 0, 0, PM_REMOVE))
+        {
+            TranslateMessage(&msg);
+            DispatchMessage(&msg);
+        }
+    }
+
+    // Unregister the hInstance so it can be reused if
+    // the sample is reloaded
+    if (!UnregisterClass(L"DXQuad", hInstance))
+    {
+        auto error = GetLastError();
+        MessageBox(NULL, L"Error unregistering class",
+            L"Error", MB_OK | MB_ICONERROR);
+        return 1;
+    }
+    else {
+        OutputDebugString(L"Class unregistered");
+    }
+
+    // Return this part of the WM_QUIT message to Windows.
+    return static_cast<char>(msg.wParam);
+}
+
+// Main message handler for the sample.
+LRESULT CALLBACK Win32Application::WindowProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
+{
+    D3D12Quad* pSample = reinterpret_cast<D3D12Quad*>(GetWindowLongPtr(hWnd, GWLP_USERDATA));
+
+    switch (message)
+    {
+    case WM_CREATE:
+    {
+        // Save the D3D12Quad* passed in to CreateWindow.
+        LPCREATESTRUCT pCreateStruct = reinterpret_cast<LPCREATESTRUCT>(lParam);
+        SetWindowLongPtr(hWnd, GWLP_USERDATA, reinterpret_cast<LONG_PTR>(pCreateStruct->lpCreateParams));
+    }
+    return 0;
+
+    case WM_PAINT:
+        if (pSample)
+        {
+            if (!close_window)
+            {
+                pSample->OnUpdate();
+                pSample->OnRender();
+            }
+        }
+        return 0;
+
+    case WM_DESTROY:
+        PostQuitMessage(0);
+        return 0;
+    }
+
+    // Handle any messages the switch statement didn't.
+    return DefWindowProc(hWnd, message, wParam, lParam);
+}
+
+void Win32Application::CloseWindow()
+{
+    close_window = true;
+}
+
+std::wstring Win32Application::GetAssetPath(LPCWSTR assetName)
+{
+    WCHAR modulePath[512];
+    UINT modulePathSize = _countof(modulePath);
+    GetModuleFileName(nullptr, modulePath, modulePathSize);
+
+    const wchar_t* gallery = L"WinMLSamplesGallery";
+    UINT gallerySize = 19; // length of the gallery string
+
+    // Erase everything after the first occurence of the gallery string
+    std::wstring path(modulePath);
+    size_t found = path.find(gallery);
```

Comment:
>size_t found = path.find(gallery); [](http://example.com/codeflow?start=4&length=34)

If we're certain this can never fail, then an assertion would be smart.

```
size_t offset = path.find(gallery);
assert(offset != std::string::npos);
```

Reply:
Done.